core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -87,7 +99,49 @@ protected boolean transportCompress() {
         return false;
     }
 
-    protected abstract NodesResponse newResponse(NodesRequest request, AtomicReferenceArray nodesResponses);
+    /**
+     * Map the responses into {@code nodeResponseClass} responses and {@link FailedNodeException}s.
+     *
+     * @param request The associated request.
+     * @param nodesResponses All node-level responses
+     * @return Never {@code null}.
+     * @throws NullPointerException if {@code nodesResponses} is {@code null}
+     * @see #newResponse(BaseNodesRequest, BaseNodeResponse[], FailedNodeException[])
+     */
+    @SuppressWarnings("unchecked") // generic array creation
```

Comment:
I think based on Nik's part, I'm going to change everything from an array to a `List`.